### ISSUE TITLE
Failure running +project/testOnly 

### ISSUE THREAD

**Issue by ashawley** [sbt/sbt#4715]
## steps

In a multi-project build for a library that cross-compiles, run a single test across multiple Scala versions:

```
> +foo/testOnly foo.BarTest
```

I'll try to send a scripted test in a PR to prove this is not working.

## problem

This used to work in sbt 0.13.  In sbt 1.2.8, there's a runtime exception:

```
[error] java.util.NoSuchElementException: key not found: ProjectRef(file:project/,root)
[error] 	at scala.collection.immutable.Map$Map1.apply(Map.scala:111)
[error] 	at sbt.Cross$.crossBuildCommandImpl(Cross.scala:167)
[error] 	at sbt.Cross$.$anonfun$crossBuild$2(Cross.scala:123)
[error] 	at sbt.Cross$$$Lambda$386/2095970915.apply(Unknown Source)
[error] 	at sbt.Command$.$anonfun$applyEffect$4(Command.scala:142)
[error] 	at sbt.Command$$$Lambda$330/509806761.apply(Unknown Source)
[error] 	at sbt.Command$.$anonfun$applyEffect$2(Command.scala:137)
[error] 	at sbt.Command$$$Lambda$356/756637412.apply(Unknown Source)
[error] 	at sbt.Command$.process(Command.scala:181)
[error] 	at sbt.MainLoop$.processCommand(MainLoop.scala:151)
[error] 	at sbt.MainLoop$.$anonfun$next$2(MainLoop.scala:139)
[error] 	at sbt.MainLoop$$$Lambda$311/1800932232.apply(Unknown Source)
[error] 	at sbt.State$$anon$1.runCmd$1(State.scala:246)
[error] 	at sbt.State$$anon$1.process(State.scala:250)
[error] 	at sbt.MainLoop$.$anonfun$next$1(MainLoop.scala:139)
[error] 	at sbt.MainLoop$$$Lambda$310/2000410491.apply(Unknown Source)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.MainLoop$.next(MainLoop.scala:139)
[error] 	at sbt.MainLoop$.run(MainLoop.scala:132)
[error] 	at sbt.MainLoop$.$anonfun$runWithNewLog$1(MainLoop.scala:110)
[error] 	at sbt.MainLoop$$$Lambda$303/1226298219.apply(Unknown Source)
[error] 	at sbt.io.Using.apply(Using.scala:22)
[error] 	at sbt.MainLoop$.runWithNewLog(MainLoop.scala:104)
[error] 	at sbt.MainLoop$.runAndClearLast(MainLoop.scala:59)
[error] 	at sbt.MainLoop$.runLoggedLoop(MainLoop.scala:44)
[error] 	at sbt.MainLoop$.runLogged(MainLoop.scala:35)
[error] 	at sbt.StandardMain$.runManaged(Main.scala:138)
[error] 	at sbt.xMain.run(Main.scala:89)
[error] 	at xsbt.boot.Launch$$anonfun$run$1.apply(Launch.scala:109)
[error] 	at xsbt.boot.Launch$.withContextLoader(Launch.scala:128)
[error] 	at xsbt.boot.Launch$.run(Launch.scala:109)
[error] 	at xsbt.boot.Launch$$anonfun$apply$1.apply(Launch.scala:35)
[error] 	at xsbt.boot.Launch$.launch(Launch.scala:117)
[error] 	at xsbt.boot.Launch$.apply(Launch.scala:18)
[error] 	at xsbt.boot.Boot$.runImpl(Boot.scala:56)
[error] 	at xsbt.boot.Boot$.main(Boot.scala:18)
[error] 	at xsbt.boot.Boot.main(Boot.scala)
[error] java.util.NoSuchElementException: key not found: ProjectRef(file:project/,root)
```

## expectation

Run the test across multiple versions of Scala, and not throw an exception.

## notes

sbt version: 1.2.8

Thanks for sbt,
Aaron

**Comment by ashawley** [comment]
Looks like this worked fine in sbt 1.0.0, but broke in sbt 1.0.1.   

A git bisect has tracked it down to be91300.

**Comment by eatkins** [comment]
@ashawley thanks for the report and the bisect sleuthing. I have confirmed the bug in a project of mine.

**Comment by ashawley** [comment]
Tried to fix in #4727, but it needs more work.

**Comment by eed3si9n** [comment]
- In sbt 0.13, `+` (cross) command would treat its argument as a "command",
and cross build based on a single set of `crossScalaVersions`.
- I added sbt-doge that changed the semantics of command aggregation by treating all of `+` argument as a task, like `compile`, and `test`, and only broadcasting `compile` etc if `crossScalaVersions` supported it - https://github.com/sbt/sbt-doge/blob/v0.1.1/src/main/scala/sbtdoge/DogePlugin.scala
- there's been subsequent improvements on that to support `+subproj/task` and `+command`

It would be good to find out what the current parser implementation thinks `foo/testOnly foo.BarTest` is.

**Comment by eed3si9n** [comment]
This is fixed in https://github.com/sbt/sbt/pull/5265

**Comment by ashawley** [comment]
Yes, looks good.  I can cross build and run `testOnly` against multiple Scala versions again.  

Thank you!

**Comment by eed3si9n** [comment]
There seems to be regression in the PR, so I am reopening this.

**Comment by ashawley** [comment]
This was reopened, but it seems multiple regressions were ironed out.  I'm looking at https://github.com/sbt/sbt/issues/5310, https://github.com/sbt/sbt/issues/5314, https://github.com/sbt/sbt/issues/5321 and https://github.com/sbt/sbt/issues/5336.  This was fixed in sbt 1.3.5 and still works in 1.3.6, 1.3.7 and 1.3.8.  It doesn't seem like this will be rolled back, so closing...
